openvdb/points/PointAdvect.h

```diff
@@ -0,0 +1,288 @@
+#include <openvdb/points/PointGroup.h>
+#include <openvdb/points/PointMove.h>
+
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace points {
+
+
+/// @brief Advect points in a PointDataGrid through a velocity grid
+/// @param points               the PointDataGrid containing the points to be advected.
+/// @param velocity             a velocity grid to be sampled.
+/// @param integrationOrder     the integration scheme to use (1 is forward euler, 4 is runge-kutta 4th)
+/// @param dt                   delta time.
+/// @param timeSteps            number of advection steps to perform.
+/// @param advectIncludeGroups  a vector of VDB Points groups to be included from advection (default is all).
+/// @param advectExcludeGroups  a vector of VDB Points groups to be excluded from advection (default is none).
+/// @param includeGroups        a vector of VDB Points groups to be included (default is all).
+/// @param excludeGroups        a vector of VDB Points groups to be excluded (default is none).
+/// @param cached               caches velocity interpolation for faster performance, disable to use
+///                             less memory (default is on).
+template <typename PointDataGridT, typename VelGridT>
+inline void advectPoints(PointDataGridT& points, const VelGridT& velocity,
+                         const Index integrationOrder, const double dt, const Index timeSteps,
+                         const std::vector<Name>& advectIncludeGroups = std::vector<Name>(),
+                         const std::vector<Name>& advectExcludeGroups = std::vector<Name>(),
+                         const std::vector<Name>& includeGroups = std::vector<Name>(),
+                         const std::vector<Name>& excludeGroups = std::vector<Name>(),
+                         const bool cached = true);
+
+
+////////////////////////////////////////
+
+
+namespace point_advect_internal {
+
+enum IntegrationOrder {
+    INTEGRATION_ORDER_FWD_EULER = 1,
+    INTEGRATION_ORDER_RK_2ND,
+    INTEGRATION_ORDER_RK_3RD,
+    INTEGRATION_ORDER_RK_4TH
+};
+
+template <typename VelGridT, Index IntegrationOrder, bool Staggered, typename FilterT>
+class AdvectionDeformer
+{
+public:
+    using IntegratorT = openvdb::tools::VelocityIntegrator<VelGridT, Staggered>;
+
+    AdvectionDeformer(const VelGridT& velocityGrid, const double timeStep, const int steps,
+                      const FilterT& filter)
+        : mIntegrator(velocityGrid)
+        , mTimeStep(timeStep)
+        , mSteps(steps)
+        , mFilter(filter) { }
+
+    template <typename LeafIterT>
+    void reset(const LeafIterT& leaf)
+    {
+        mFilter.reset(*leaf);
+    }
+
+    template <typename IndexIterT>
+    void apply(Vec3d& position, const IndexIterT& iter) const
+    {
+        if (mFilter.valid(iter)) {
+            for (int n = 0; n < mSteps; ++n) {
+                mIntegrator.template rungeKutta<IntegrationOrder, openvdb::Vec3d>(mTimeStep, position);
```

Comment:
Looks like mTimeStep never gets divided by mSteps prior to this call so is it currently stepping too far if there are multiple advection substeps?

Reply:
@dnegmw I agree this feels a little counter-intuitive, but in fact the timestep parameter here is already per-step. Here's where the division you're referring to happens in the SOP:

https://github.com/dreamworksanimation/openvdb/blob/master/openvdb_houdini/houdini/SOP_OpenVDB_Advect_Points.cc#L1021

If you pass in Houdini points and VDB points while varying these parameters, you should see identical behaviour.